hyperstart/api/grpc/utils.go

```diff
@@ -0,0 +1,140 @@
+package grpc
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func copyValue(to, from reflect.Value) error {
+	toKind := to.Kind()
+	fromKind := from.Kind()
+
+	if !from.IsValid() {
+		return nil
+	}
+
+	if toKind == reflect.Ptr {
+		// If the destination is a pointer, we need to allocate a new one.
+		to.Set(reflect.New(to.Type().Elem()))
+		if fromKind == reflect.Ptr {
+			return copyValue(to.Elem(), from.Elem())
+		} else {
+			return copyValue(to.Elem(), from)
+		}
+	} else {
+		// Here the destination is not a pointer.
+		// Let's check what's the origin.
+		if fromKind == reflect.Ptr {
+			return copyValue(to, from.Elem())
+		}
+
+		switch toKind {
+		case reflect.Struct:
+			return copyStructValue(to, from)
+		case reflect.Slice:
+			return copySliceValue(to, from)
+		case reflect.Map:
+			return copyMapValue(to, from)
+		default:
+			// We now are copying non pointers scalar.
+			// This is the leaf of the recursion.
+			if from.Type() != to.Type() && from.Type().ConvertibleTo(to.Type()) {
+				to.Set(from.Convert(to.Type()))
+				return nil
+			} else {
+				to.Set(from)
+				return nil
+			}
+
+		}
+	}
+}
+
+func copyMapValue(to, from reflect.Value) error {
+	if to.Kind() != reflect.Map && from.Kind() != reflect.Map {
+		return fmt.Errorf("Can only copy maps into maps")
+	}
+
+	mapLen := from.Len()
+	to.Set(reflect.MakeMap(to.Type()))
+
+	for j := 0; j < mapLen; j++ {
+		if err := copyValue(to.Index(j), from.Index(j)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func copySliceValue(to, from reflect.Value) error {
+	if to.Kind() != reflect.Slice && from.Kind() != reflect.Slice {
+		return fmt.Errorf("Can only copy slices into slices")
+	}
+
+	sliceLen := from.Len()
+	to.Set(reflect.MakeSlice(to.Type(), sliceLen, sliceLen))
+
+	for j := 0; j < sliceLen; j++ {
+		if err := copyValue(to.Index(j), from.Index(j)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func copyStructValue(to, from reflect.Value) error {
+	if to.Kind() != reflect.Struct && from.Kind() != reflect.Struct {
+		return fmt.Errorf("Can only copy structs into structs")
+	}
+
+	if to.NumField() != from.NumField() {
+		return fmt.Errorf("Structures must have the same number of fields")
+	}
+
+	for i := 0; i < to.NumField(); i++ {
+		if err := copyValue(to.Field(i), from.Field(i)); err != nil {
```

Comment:
I think it should be copied based on the filed name instead of the index.

Reply:
@laijs I thought about that option but I'd rather avoid it due to this documented bug:
https://golang.org/pkg/reflect/#pkg-note-BUG

Reply:
Since they can be decoded from json string, which conversion is based on the filed names/tags. I think the problem can be resolved here. (https://github.com/golang/go/blob/a3e013b0824ba53168b5d91abdb6ce191510a89d/src/encoding/json/encode.go#L1231)

I'm worried when the oci spec is changed in future, how would we maintain the compatibilities. (vs. older gRPC, vs. older OCI, and vs. newer OCI)? I seams that copying manually field by field is good. We can use this code, because we can change it to "manually copying" any time when any problem happen, or to a new method any time when we found one.

Reply:
@laijs I'm not sure if you're asking for a change here or not. Implementing something similar to the JSON implementation means we would have to maintain additional metadata for all our fields, sort them through a similar function as well, etc...I believe this is an overkill.
I can add an additional check to verify that we're copying fields with the same name and error out when/if we're not. Any OCI config incompatibility would then be caught at unit testing time.

Reply:
@laijs I changed the logic in `copyStructValue` to verify that:
* We copy from fields with the same name
* If fields don't have the same name, then we go through a loop searching for the right field in the from structure.

Reply:
@laijs I think having this full OCI specification structure will be a pain to maintain. Don't you think that having only a `string` field would be easier ? I mean we could pass the entire config.json file serialized as a JSON and we would unmarshal from the agent. That way, we just need to make sure that both runtime and agent rely on the same OCI spec version. I don't see any benefits in having this described into the gRPC protocol description here.